rusoto/services/timestream-query/src/custom/mod.rs

```diff
@@ -0,0 +1,158 @@
+                    RusotoError::Service(DescribeEndpointsError::InternalServer(
+                        "DescribeEndpoints API returned empty endpoint list".to_string(),
+                    )),
+                )?;
+
+                self.endpoint = Some(Endpoint {
+                    address: endpoint.address.clone(),
+                    // Subtract 1 minute from the TTL to ensure the endpoint data is refreshed before it expires.
+                    expiry: Instant::now().add(Duration::from_secs(max(
+                        5,
+                        endpoint.cache_period_in_minutes as u64 * 60 - 60,
+                    ))),
+                });
+                self.client = TimestreamQueryClient::new(Region::Custom {
+                    name: self.region.name().to_string(),
+                    endpoint: endpoint.address,
+                });
+
+                Ok(())
+            }
+        }
+    }
+
+    /// <p> Cancels a query that has been issued. Cancellation is guaranteed only if the query has not completed execution before the cancellation request was issued. Because cancellation is an idempotent operation, subsequent cancellation requests will return a <code>CancellationMessage</code>, indicating that the query has already been canceled. </p>
+    pub async fn cancel_query(
+        &mut self,
+        input: CancelQueryRequest,
+    ) -> Result<CancelQueryResponse, RusotoError<CancelQueryError>> {
+        self.update_endpoint().await.map_err(|e| match e {
+            // Remap any DescribeEndpointError to ThrottlingErrors
+            RusotoError::Service(DescribeEndpointsError::Throttling(s)) => {
+                RusotoError::Service(CancelQueryError::Throttling(s))
+            }
+            RusotoError::Service(DescribeEndpointsError::InternalServer(s)) => {
+                RusotoError::Service(CancelQueryError::InternalServer(s))
+            }
+            // Pass on all non-service related errors as is.
+            RusotoError::Blocking => RusotoError::Blocking,
+            RusotoError::Credentials(e) => RusotoError::Credentials(e),
+            RusotoError::HttpDispatch(e) => RusotoError::HttpDispatch(e),
+            RusotoError::ParseError(e) => RusotoError::ParseError(e),
+            RusotoError::Unknown(e) => RusotoError::Unknown(e),
+            RusotoError::Validation(e) => RusotoError::Validation(e),
+        })?;
+
+        self.client.cancel_query(input).await
+    }
+
+    /// <p> Query is a synchronous operation that enables you to execute a query. Query will timeout after 60 seconds. You must update the default timeout in the SDK to support a timeout of 60 seconds. The result set will be truncated to 1MB. Service quotas apply. For more information, see Quotas in the Timestream Developer Guide. </p>
+    pub async fn query(
+        &mut self,
+        input: QueryRequest,
+    ) -> Result<QueryResponse, RusotoError<QueryError>> {
+        self.update_endpoint().await.map_err(|e| match e {
```

Comment:
may be misunderstanding this, but does this mean we update the endpoint every time we do a call?

Reply:
Yes, this is due to the endpoints for queries are dynamic. The same for the writer. Also see here [service.json of boto](https://github.com/boto/botocore/blob/23ee17f5446c78167ff442302471f9928c3b4b7c/botocore/data/timestream-query/2018-11-01/service-2.json#L33)

Reply:
Yes, the endpoints are dynamic -- but they are meant to be cached for the TTL that comes with them (i.e point 3 here:https://docs.aws.amazon.com/timestream/latest/developerguide/Using-API.endpoint-discovery.describe-endpoints.implementation.html). On a second pass though, we _don't_ actually update the endpoint with each call, only when we find at the time of calling that the expiry for the endpoint has passed. Which seems correct.

Reply:
So this is ok then?